ssl/ssl_local.h

```diff
@@ -2446,11 +2467,47 @@ struct openssl_ssl_test_functions {
 
 const char *ssl_protocol_to_string(int version);
 
+static ossl_inline int tls12_rpk_and_privkey(const SSL_CONNECTION *sc, int idx)
+{
+    /*
+     * This is to check for special cases when using RPK with just
+     * a private key, and NO CERTIFICATE
+     */
+    return ((sc->server && sc->ext.server_cert_type == TLSEXT_cert_type_rpk)
+            || (!sc->server && sc->ext.client_cert_type == TLSEXT_cert_type_rpk))
+        && sc->cert->pkeys[idx].privatekey != NULL
+        && sc->cert->pkeys[idx].x509 == NULL;
```

Comment:
@mattcaswell What's your view on ensuring conformant point compression (typically lack thereof)?

Reply:
Is non-compressed an RPK requirement?

Reply:
Not as far as I can tell. It might be more of an OpenSSL limitation?

Reply:
Hmm. With a normal certificate we run various validity checks as part of tls1_set_cert_validity->tls1_check_chain->tls1_check_cert_param. In particular `tls1_check_cert_param` checks that the pkey conforms with the point compression requirements.

With RPK, if the server has been configured with a certificate but it isn't used because RPK is negotiated, it seems the pkey is still checked to ensure the point compression requirements are met.

However if the server has not been configured with a certificate at all tls1_check_chain finishes early (silently returning an error code) and the point compression checks are therefore skipped.

This doesn't seem correct.

Reply:
(I hate that github isn't showing this as a comment thread...)

In `tls1_check_chain()`, there's some early checks for the presence of both a cert and a key, and if either is missing, the function exits early. The point compression checks are skipped, but effectively they fail because this function `tls_check_chain()` fails. So, this is potentially a false negative for RPK.

The trick here is that the certificate isn't valid; but the RPK might be. Let me think about this a bit, but I welcome ideas.

Reply:
I added a new `PKEY_CERT_RPK` flag to "validate" the CERT_PKEY entry as being valid for RPK use (i.e. `tls1_check_pkey_comp()`). Then I used this flag to set the `k` and `a` flags in `ssl_set_masks()`. Removed DSA and GOST support while I was in there.

Reply:
So, @mattcaswell, I think this comment is addressed.